keen/keen-js
# Fix 'getSetupTemplate' so it accepts some custom user configuration in 'chartOptions'

## What does this PR do?

When a user tries to set axis ticks with 'chartOptions', their configuration gets overwritten by defaults.
Example:
```javascript
var dv = new Dataviz()
dv.adapter('c3')
.chartOptions({
  axis: { x: { ticks: 2 } }
}) // this gets overwritten
```

This PR lets the user overwrite `x` axis configurations with their own in `chartOptions`. Additionally, to test this change, `getSetupTemplate` (the code that creates the configuration object that is passed to C3) is pulled into it's own module so it can be easily unit tested.

## How to test this PR:

* Look at new tests for `getSetupTemplate`
* Run all tests, make sure they pass
* Open `test/examples/dataviz/c3/index.html` and make sure the `Multi-group_by-interval (Custom axis)` example works.

## Caveats:

Configuring the x axis WILL overwrite some defaults in keen-js. So if you have a chronological chart that looks like this: 

![screen shot 2016-04-12 at 10 48 32 am](https://cloud.githubusercontent.com/assets/638351/14468945/701c6518-009f-11e6-876a-9c015f502f1a.png)

And you pass in `{ axis: { x: { ticks: 2 } } }` to `chartOptions`, your chart will now look like this:
![screen shot 2016-04-12 at 10 48 49 am](https://cloud.githubusercontent.com/assets/638351/14468960/8a52acc6-009f-11e6-81a2-e89f01daad17.png)

To solve this, pass in an optional `format`, like so (see another example here: http://c3js.org/samples/axes_x_tick_format.html)
```javascript
{  axis: { x: { format: '%m/%Y', ticks: 2 } } }
```

![screen shot 2016-04-12 at 10 54 42 am](https://cloud.githubusercontent.com/assets/638351/14469011/c81fedde-009f-11e6-8dc4-c455db2e06ab.png)
